[azurerm_batch_pool] - support for custom images with the `storage_image_reference` property

jcorioland opened the pull request:
This PR brings support to custom image for Azure Batch Pool.
It fixes #3352.

- [x] Implementation in resource
- [x] Implementation in data source
- [x] Acceptance tests
- [x] Documentation + Example

jcorioland commented:
The custom image id need to be set in the `id` property of the storage_image_reference in the API. I am wondering if it's better to do it this way (like it is currently):

```
storage_image_reference {
    id = "${data.azurerm_image.test.id}"
}
```

or add an extra property to the pool, which is not nested to the image reference (like proposed in the issue) and do the assignation in the code.

@tombuildsstuff @katbyte any thoughts?

jeffreyCline commented on website/docs/r/batch_pool.html.markdown:
````diff
@@ -120,6 +120,21 @@ The following arguments are supported:
 
+~> **NOTE:** It's possible to reference a custom VM image for the pool by specifying it's `id` in the `id` property of the `storage_image_reference`. This property is mutually exclusive with other properties. The VM image should be in the same region that the batch pool you are about to create. See [official documentation](https://docs.microsoft.com/en-us/azure/batch/batch-custom-images) for more ...
+
+```
````
Shouldn't this code sample be moved to a new section under the **Example Usage** section? Say something like ## Example Usage with Custom Image?

jcorioland replied:
Yes, totally agree.

jcorioland commented:
@jeffreyCline thank you for the review!  I've made some updates according to your comment.

jcorioland commented:
Hey @jeffreyCline @tombuildsstuff @katbyte - any more change required on this one?

jcorioland commented:
@katbyte - the tests are now fixed :) the travis ci fails because of spell checking failing on hdinsight, but nothing that seems related to this PR.

jcorioland commented:
Hi @katbyte - is it anything that you're waiting from me on this PR?

katbyte commented on website/docs/r/batch_pool.html.markdown:
````diff
@@ -102,6 +102,35 @@ EOF
 }
 ```
 
+## Example Usage with a Custom VM Image
````
Could we move this to the example folder: `examples/batch/pool/custom-image` or such?

jcorioland replied:
done.

katbyte commented on website/docs/r/batch_pool.html.markdown:
```diff
@@ -120,6 +149,8 @@ The following arguments are supported:
 
 * `storage_image_reference` - (Required) A `storage_image_reference` for the virtual machines that will compose the Batch pool.
```
I don't see any docs for the block here, could we add that in?

jcorioland replied:
added.

katbyte commented on website/docs/r/batch_pool.html.markdown:
```diff
@@ -120,6 +149,8 @@ The following arguments are supported:
 
 * `storage_image_reference` - (Required) A `storage_image_reference` for the virtual machines that will compose the Batch pool.
 
+~> **NOTE:** It's possible to reference a custom VM image for the pool by specifying it's `id` in the `id` property of the `storage_image_reference`. This property is mutually exclusive with other properties. The VM image should be in the same region that the batch pool you are about to create. See [official documentation](https://docs.microsoft.com/en-us/azure/batch/batch-custom-images) for more ...
```
This note should be on the id property documentation.

jcorioland replied:
done.

katbyte commented on examples/batch/main.tf:
```diff
@@ -83,3 +83,26 @@ EOF
   }
 }
+
+data "azurerm_image" "image" {
```
Could we split this out into another folder with a complete working example that doesn't depend on an existing image?

jcorioland replied:
Done.

katbyte commented on azurerm/resource_arm_batch_pool.go:
```diff
@@ -137,34 +137,35 @@ func resourceArmBatchPool() *schema.Resource {
 						"id": {
 							Type:         schema.TypeString,
 							Optional:     true,
+							ForceNew:     true,
```
Whats the reasoning behind this change? it looks like this property was here but ignored before?

jcorioland replied:
Actually it was ignored because the Batch Pool provider did not support to provide the id of a custom image (which is what this PR brings). The `id` field is only used in that case. I also updated all the other field to be optional as `id` is mutually exclusive with the four other properties.

jcorioland commented:
@katbyte thank you for the review! I've fixed the issue in the documentation and also split the example into a `basic` and `custom-img` folder to make sure we cover both use cases separately. I also made sure that the custom image example is not dependant from an existing custom image but creates it itself.
All the changes you've requested should be covered now. LMK if there is anything else I need to improve.
Thanks!
